12.0 add commown devices send modules from crm lead

9184Lucky opened the pull request.

Commit:
Create functions to send lots and secondary products on the same picking

Commit:
Modify crm.lead picking wizard to send secondary products

Commit:
Add model describing relations between service variants and storable products

Commit:
Display service_storable_config on product template

Added service storable configuration field on product template and a view to display them.
This view allows configurations edition directly on product template
view.

Commit:
Add fields storing primary & secondary storable product on service variant

Also added a funnction to compute those fields using Service storable
configuarions.

fcayre commented on commown_devices/models/product_service_config.py:
```diff
@@ -0,0 +1,41 @@
+
+    service_tmpl_id = fields.Many2one(
+        "product.template",
+        required=True,
```
`string="Configured product template",`

9184Lucky replied:
Fait

fcayre commented on commown_devices/models/product_service_config.py:
```diff
@@ -0,0 +1,41 @@
+            ("primary", "Primary product"),
+            ("secondary", "Secondary product"),
+        ],
+        string="Configuration type",
```
```
string="Configured storable field",
help="Does current config concern the primary storable or the secondary storables of current product variants?",
```
Peut-être renommer le champ en `storable_type` ?

9184Lucky replied:
Rajout fait et oui pour le renomage

fcayre commented on commown_devices/models/product.py:
```diff
@@ -15,3 +15,42 @@ class Product(models.Model):
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    primary_variant_id = fields.Many2one(
```
`primary_storable_variant_id` plutôt non ? Même proposition pour secondary

9184Lucky replied:
Fait

fcayre commented on commown_devices/models/product.py:
```diff
@@ -15,3 +15,42 @@ class Product(models.Model):
+                lambda c: set(c.attribute_value_ids.ids).issubset(
+                    rec.attribute_value_ids.ids
+                )
+            )
```
Trier les configs en une seule boucle + vérifier qu'on a bien une config primary max qui s'applique.

9184Lucky replied:
Ok, fait

fcayre commented on commown_devices/models/common.py:
```diff
@@ -60,6 +60,176 @@ def internal_picking_tracking_none(
 
 
+def first_common_location(locs):
+    env = locs[0].env
```
Ajouter une docstring

9184Lucky replied:
Fait

fcayre commented on commown_devices/models/common.py:
```diff
@@ -60,6 +60,176 @@ def internal_picking_tracking_none(
+                ]
+            )
+            .filtered(
+                lambda q, qts=quantity_to_send: q.quantity - q.reserved_quantity > qts
```
`>=` non ?

9184Lucky replied:
Oui carrément !

fcayre commented on commown_devices/models/common.py:
```diff
@@ -60,6 +60,176 @@ def internal_picking_tracking_none(
+    return pts_orig
+
+
+def find_lots_orig_location(env, lots):
```
Je pense que ce helper pourrait être utilisé partout où on utilise une `lambda q: q.quantity` > 0 à condition que : 
- il prenne un lot unique comme argument (il est alors facile de faire un {lot: current_location(lot) for lot in lots} pour obtenir le même résultat que ta fonction à lots multiples) et retourne un emplacement
- il prenne un argument `search_in` pour spécifier un emplacement parent
- il prenne un argument `raise_if_not_found=False` pour savoir si on raise quand on ne le trouve pas
- il prenne un argument `raise_if_reserved=False` pour savoir si on raise quand il est réservé

On pourrait même en faire une méthode de stock.production.lot (pourquoi ça existe pas ?!) non ?

fcayre commented on commown_devices/models/common.py:
```diff
@@ -60,6 +60,176 @@ def internal_picking_tracking_none(
+
+
+def find_products_orig_location(env, products):
+    """From a dictionary {product: qantity}
```
qUantity

fcayre commented on commown_devices/models/common.py:
```diff
@@ -60,6 +60,176 @@ def internal_picking_tracking_none(
+def find_products_orig_location(env, products):
+    """From a dictionary {product: qantity}
+    find the location from where product
+    can be send"""
```
senT

fcayre commented on commown_devices/models/common.py:
```diff
@@ -60,6 +60,176 @@ def internal_picking_tracking_none(
+    """From a dictionary {product: qantity}
+    find the location from where product
+    can be send"""
+    stock = env.ref("stock.stock_location_stock")
```
J'ai l'impression que ça devrait être un argument facultatif de la méthode (genre `stock=None`) en disant que s'il n'est pas fourni on prend `stock.stock_location_stock`.

fcayre commented on commown_devices/models/common.py:
```diff
@@ -60,6 +60,176 @@ def internal_picking_tracking_none(
+    return picking
+
+
 def internal_picking(
```
Ne doit-on pas supprimer cette méthode au profit de la tienne ? (du coup on peut garder le même nom)

9184Lucky replied:
Oui, le projet à la fin est de faire un refactoring qui suprime internal_picking et internal picking_picking_tracking_none au profit de la nouvelle fonction (ainsi que send_device et send_devices_tracing_none du coup)

fcayre commented on commown_devices/models/contract.py:
```diff
@@ -95,6 +99,25 @@ def send_device_tracking_none(
             do_transfer=do_transfer,
         )
 
+    def send_devices_mixt(
```
Idem : est-ce que ça ne doit pas devenir la nouvelle méthode `send_devices` ?

9184Lucky replied:
Si carrément, c'est dans les plans !

fcayre commented on commown_devices/models/product_service_config.py:
```diff
@@ -0,0 +1,41 @@
+    attribute_value_ids = fields.Many2many(
+        "product.attribute.value",
+        string="Restrict to variants with following attribute values",
+        required=False,
```
Restreindre le domain aux attribute values du modèle d'article.

fcayre commented on commown_devices/models/product_service_config.py:
```diff
@@ -0,0 +1,41 @@
+        "product.product",
+        required=True,
+        domain=(
+            "[('type', '=', 'product')," "('product_tmpl_id', '=', storable_tmpl_id)]"
```
enlever les guillemets inutiles au milieu.

Commit:
i18n fr10n

fcayre commented on commown_devices/models/stock_production_lot.py:
```diff
@@ -12,3 +12,30 @@ def name_get(self):
             result.append((record.id, name))
         return result
+
+    def current_location(
```
Vraiment proche de `find_products_orig_location`, pas moyen de factoriser ?
Pas sûr que `find_products_orig_location` soit vraiment utile.

9184Lucky replied:
`find_products_orig_location` a été crée pour un cas plus générique où on envoi plus d'une unitée et produits et où les stocks de modules sont répartis dans différents emplacements... en l'état pas bien utile mais comme je compte unifier les fonctions de picking je trouvais ça interessant de rester assez général et d'envisager ce genre de cas. A noter que on aurait pu faire encore mieux et permettre de prendre une certaine quantitée dans un emplacement et de completer avec d'autre emplacements (Ex le cas ou on veut envoyer du neuf et du reconditionné).
Même si y'a beaucoup de similarité avec current_location et donc peut être possibilité de factoriser il y a pour moi deux différences majeures avec current_location. La première est que on cherche des entitées différentes (product.product et stock.production.lot) et qu'il en découle que les notions de quantité ne s'appliquent pas aux stock_production_lot qui est forcément unique..

9184Lucky replied:
Cependant j'ai peut être compliqué le code pour rien et on peut se dire que dans l'état c'est pas necessaire quitte à rajouter cette fonction le jour où le besoin s'en fait sentir....

fcayre commented on commown_devices/tests/test_common.py:
```diff
@@ -0,0 +1,65 @@
+from ..models.common import first_common_location
+
+
+def create_stock_location(self, name, parent=None):
```
Trop de copier-coller ci-dessous. Écrire un truc genre : 
```
return self.env["stock.location"].create(
    {
                "name": name,
                "usage": "internal",
                "partner_id": 1,
                "location_id": parent.id if parent else False,
    }
)
```

9184Lucky replied:
Fait

fcayre commented on commown_devices/tests/test_common.py:
```diff
@@ -0,0 +1,65 @@
+        )
+
+
+class CommonFunctionsTC(TransactionCase):
```
Utiliser plutôt `SavepointCase` comme base, beaucoup plus rapide quand on ajoute des tests (savepoint à la fin du setup => on ne ré-éxecute pas le setup à chaque test).

9184Lucky replied:
Fait

fcayre commented on commown_devices/models/product.py:
```diff
@@ -15,3 +15,47 @@ class Product(models.Model):
+    )
+
+    @api.multi
+    def _set_storable_variants(self):
```
Il manque un cas d'appel très important à cette méthode : quand on modifie les attribute_value_ids d'un product template et ainsi ses variantes.

9184Lucky replied:
Bien vu ! Et fait.

fcayre reviewed (changes requested):
__BILAN de la relecture__

Une seule demande vraiment importante : il manque le cas le plus important d'appel à la configuration des stockables sur un service, quand on modifie ses variantes (typiquement : ajout d'une couleur qu'on vient de restocker).

Autres remarques non notées ailleurs : 
- wizard ne prend pas deux z
- l'idéal est de ne pas coder tous les tests à la fin, mais en cohérence avec les ajouts de code ; ça permet d'un coup d'oeil de voir que la couverture de test suffit (et pas seulement la couverture technique, la couverture logique aussi), et ça rend (encore !), la relecture approfondie plus facile.

fcayre commented:
**BILAN de la relecture**

Une seule demande vraiment importante : il manque le cas le plus important d'appel à la configuration des stockables sur un service, quand on modifie ses variantes (typiquement : ajout d'une couleur qu'on vient de restocker).

Autres remarques non notées ailleurs :

-  wizard ne prend pas deux z
-  l'idéal est de ne pas coder tous les tests à la fin, mais en cohérence avec les ajouts de code ; ça permet d'un coup d'oeil de voir que la couverture de test suffit (et pas seulement la couverture technique, la couverture logique aussi), et ça rend (encore !), la relecture approfondie plus facile.

9184Lucky commented:
Salut Flo, c'est corrigé en tenant compte de tes remarques... J'essaierais de faire les test plus au fur et à mesure la prochaine fois !
Juste sur la fonction `find_product_orig_location`que j'ai pas appliqué de correctif, à ta dispo pour en discuter.

Commit:
Rewrite code tu use more generic picking function.

Merge all picking function in a more generic one, modifiy send and
receive function of contracts accordingly.

Commit:
Replace quant by lot in function send_device of contract model

Doing so makes it more coherent with function receive_device and also
link the contract to lot more than quant wich is needed in the next
refectoring.

Commit:
Unify send contract send_device function

Commit:
Unify receive_device function of contract model

Commit:
Allow user to choose if the non serial products have to be sent from New stock or repackaged

This is useless for the lots since their localisation is unique and
found from the quants.

Commit:
Add stock location for repackaged nonserial products

Commit:
Add a picking wizard for non serial product on project task

fcayre commented on commown_devices/data/stock_location.xml:
```diff
@@ -60,6 +67,7 @@
       <field name="name">Location for devices awaiting a contract transfer</field>
       <field name="usage">internal</field>
       <field name="partner_id">1</field>
+      <field name="location_id" ref="stock_location_available_for_rent"/>
```
We should not do that.

9184Lucky replied:
Retiré, c'était même plus utile...

fcayre commented on commown_devices/models/common.py:
```diff
@@ -1,93 +1,136 @@
+                break
+        if not quants:
+            raise UserError(
+                _("Not enough %s under location %s") % (product.name, stock.name)
```
stock.name > ", ".join(stocks.mapped("name"))

9184Lucky replied:
fait

fcayre commented on commown_devices/models/common.py:
```diff
@@ -1,93 +1,136 @@
+            )
+            if quants:
+                break
+        if not quants:
```
else sur la boucle for évite ce test sur quants

9184Lucky replied:
fait

fcayre commented on commown_devices/models/common.py:
```diff
@@ -1,93 +1,136 @@
+    sent and produce a summary to tell user where each product should be sent from.
+    `stocks` parameter is a list of stock.location to look in for `products` ordered by
+    preference"""
+    if stocks == None:
```
if stocks is None

9184Lucky replied:
fait

fcayre commented on commown_devices/models/common.py:
```diff
@@ -1,93 +1,136 @@
+                    ]
+                )
+                .filtered(
+                    lambda q, qts=quantity_to_send: q.quantity - q.reserved_quantity
```
cette lambda ne depend pas de la boucle sur les stocks > la nommer et la sortir.
Éventuellement la sortir des 2 boucles et utiliser un functools.partial pour spécifier qts.

9184Lucky replied:
fait

fcayre commented on commown_devices/models/product.py:
```diff
@@ -1,11 +1,80 @@
+        for rec in self:
+            if template_configs is None:
+                template_configs = rec.product_tmpl_id.storable_config_ids
+            configs = template_configs.filtered(
```
Ajouter un commentaire ici, car c'est une ligne essentielle à comprendre. Genre : 
`# Select applicable product service configs: those which attribute values match all current product attribute values`
(ou un truc comme ça)

9184Lucky replied:
Fait

fcayre commented on commown_devices/models/product.py:
```diff
@@ -1,11 +1,80 @@
+                    % rec.name
+                )
+            rec.primary_storable_variant_id = primary_variant
+            rec.secondary_storable_variant_ids = secondary_variants
```
Duh, ça marche ça ? J'aurais pensé qu'un `rec.update({"secondary_storable_variant_id": [(6, 0, secondary_variants.ids]})` était nécessaire. Et tant qu'à faire ajouter dans cet update la mise à jour de l'attribut primary_storable_variant_id.

9184Lucky replied:
Haha alors oui ça marche ^^ 
Et oui update est plus litéral et concis vu que on peut faire les deux en même temps... Donc +1 pour l'utiliser.
Par contre il me semble que la syntaxe [(6, 0, secondary_variants.ids] n'est pas obligatoire (ça fonctionne sans) et si ça se vérifie je préfère ne pas l'utiliser je la trouve lourde et pas claire...

fcayre commented on commown_devices/models/stock_production_lot.py:
```diff
@@ -12,3 +12,30 @@ def name_get(self):
+        """Search if a lot if present in a stock location or its children and return the location"""
+
+        searched_stock_locations = self.env["stock.location"].search(
+            [
```
Le domaine doit tenir sur une seule ligne, ce serait plus lisible.

9184Lucky replied:
Carrément

fcayre commented on commown_devices/models/stock_production_lot.py:
```diff
@@ -12,3 +12,30 @@ def name_get(self):
+            ]
+        )
+
+        quant = self.quant_ids.filtered(lambda q: q.quantity > 0)
```
Une erreur explicite quand len(quant) > 1 serait sans doute mieux (plus compréhensible pour l'utilisateur final).

fcayre commented on commown_devices/models/wizard_crm_lead_picking.py:
```diff
@@ -16,55 +18,106 @@ class CrmLeadPickingWizard(models.TransientModel):
+        for rec in self:
+            picked_product_ids = rec.lot_ids.mapped("product_id").ids
+            ids_to_include = rec.all_products.filtered(
+                lambda p, picked_product_ids=picked_product_ids: p.tracking == "serial"
```
Je te vois beaucoup écrire les lambda en passant comme argument la variable de contexte que tu veux utiliser à l'intérieur de la lambda. Ce n'est pas nécessaire, et ça alourdit le code, je suis donc pour arrêter cette pratique.

9184Lucky replied:
+1

fcayre commented on commown_devices/tests/test_product.py:
```diff
@@ -0,0 +1,212 @@
+            ]
+        )
+
+    def test_behavior_on_new_variant(self):
```
Changer le nom de la méthode ou ajouter une docstring (affichée en cas de plantage du test), pour qu'on comprenne au 1er coup d'oeil ce qui est testé. Exemple : 
`"""Product service configs must be set on new product variants"""`

9184Lucky replied:
fait

fcayre commented on commown_devices/tests/test_product.py:
```diff
@@ -0,0 +1,212 @@
+            self.protective_screen.product_variant_id,
+        )
+
+    def test_behavior_on_new_config(self):
```
Idem

9184Lucky replied:
fait

fcayre commented on commown_devices/tests/test_product.py:
```diff
@@ -0,0 +1,212 @@
+        )
+
+    def test_error_on_multiple_primary_config(self):
+        # Check that it is impossible to have more than one primary variant associted to one service
```
Change the comment into a docstring

9184Lucky replied:
Fait

fcayre commented on commown_devices/tests/test_product.py:
```diff
@@ -0,0 +1,212 @@
+        )
+
+    def test_error_on_multiple_secondary_config(self):
+        """Check that a warning is issued if two secondary variants have the same template"""
```
:+1:

fcayre commented on commown_devices/tests/test_product.py:
```diff
@@ -0,0 +1,212 @@
+                % self.fp3_service_color2.name
+            ),
+            err2.exception.name,
+        )
```
Il manque des tests sur la création, la mise à jour et l'effacement de product service config. Ils sont faciles à écrire et amélioreront la couverture.

fcayre reviewed (changes requested):
Autant prendre en compte mes remarques avant de mettre en prod, mais on est vraiment proche ici !

Commit:
i18n fr.po

fcayre commented:
/ocabot merge patch
